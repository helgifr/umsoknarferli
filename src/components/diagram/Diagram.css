.diagram-wrapper {
  width: 100%;
  height: 100%;
}

.diagram__no-data {
  width: 100%;
  height: 100%;
  margin: auto;
  text-align: center;
  display: flex;
  vertical-align: middle;
  flex-direction: column;
  justify-content: center;
}