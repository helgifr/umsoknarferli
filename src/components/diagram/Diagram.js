import React, { Component } from 'react';

import { ResponsiveSankey } from '@nivo/sankey';
import { calcSankey } from '../../utils/calc';

import 'react-date-range/dist/styles.css';
import 'react-date-range/dist/theme/default.css';

import { DateRangePicker } from 'react-date-range';

import Loading from '../../components/loading';

import './Diagram.css';

class Diagram extends Component {

  state = {
    selectionRange: null,
    minDate: null,
    maxDate: null,
  }

  componentDidMount() {
    const { data } = this.props;
    const dates = data.map(el => new Date(el.date));
    const minDate = new Date(Math.min(...dates));
    const maxDate = new Date(Math.max(...dates));
    const selectionRange = {
      startDate: minDate,
      endDate: maxDate,
      key: 'selectionRange',
    }
    this.setState({ minDate, maxDate, selectionRange });
  }

  handler = (ranges) => {
    const { selectionRange } = ranges;
    this.setState({ selectionRange });
  }

  noApplications(selectionRange) {
    return (
      <div className="diagram-wrapper">
        <h2 className="diagram__no-data">Engar umsóknir</h2>
        {this.dateRangePicker(selectionRange)}
      </div>
    );
  }

  dateRangePicker(selectionRange) {
    return (
        <DateRangePicker
            ranges={[selectionRange]}
            minDate={new Date(this.state.minDate)}
            maxDate={new Date(this.state.maxDate)}
            onChange={this.handler}
            months={2}
            direction="horizontal"
          />
    );
  }

  render() {
    const { selectionRange } = this.state;

    if (!selectionRange) return (<Loading />);

    const { data } = this.props;
    const { startDate, endDate } = selectionRange;

    const filteredData = data.filter(el => new Date(el.date) >= startDate && new Date(el.date) <= endDate);

    if (filteredData.length === 0) return this.noApplications(selectionRange);

    const { nodes, links } = calcSankey(filteredData);

    return (
      <div className="diagram-wrapper">
        <ResponsiveSankey
          data={{ nodes, links }}
          margin={{
            "top": 40,
            "right": 160,
            "bottom": 40,
            "left": 50
          }}
          align="justify"
          colors="category10"
          nodeOpacity={1}
          nodeWidth={18}
          nodeBorderColor="inherit:darker(0.8)"
          linkHoverOthersOpacity={0.1}
          labelPosition="outside"
          labelOrientation="vertical"
          labelPadding={16}
          labelTextColor="inherit:darker(1)"
          enableLabels
          labelFormat={(id) => { // Skilar label í stað id
            const { label } = nodes.find(el => el.id === id);
            return label;
          }}
          motionStiffness={120}
          motionDamping={11}
        />
        {this.dateRangePicker(selectionRange)}
      </div>
    );
  }
}

export default Diagram;