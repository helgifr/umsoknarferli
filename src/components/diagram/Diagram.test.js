import React from 'react';
import { shallow } from 'enzyme';

import Diagram from './Diagram';

const data = [
  {
      "company": "Vettvangur",
      "posName": "Vefforritari",
      "date": "2018-11-10",
      "process": [
          "Bið athugun"
      ]
  },
  {
      "company": "Advania",
      "posName": "Framendaforritari",
      "date": "2018-11-16",
      "process": [
          "Viðtal",
          "Annað viðtal",
          "Neitun"
      ]
  },
  {
      "company": "Origo",
      "posName": "Vefforritari",
      "date": "2018-11-16",
      "process": [
          "Viðtal"
      ]
  }
];

describe('Diagram component', () => {
  it('matches the snapshot without data', () => {
    const component = shallow(<Diagram />);
    expect(component).toMatchSnapshot();
  });

  it('matches the snapshot with data', () => {
    const component = shallow(<Diagram data={data} />);
    expect(component).toMatchSnapshot();
  });
});