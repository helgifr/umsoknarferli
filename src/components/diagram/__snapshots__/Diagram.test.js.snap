// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagram component matches the snapshot with data 1`] = `
<div
  className="diagram-component"
>
  <div
    className="diagram-wrapper"
  >
    <ResponsiveSankey
      align="justify"
      colors="category10"
      data={
        Object {
          "links": Array [
            Object {
              "source": "Vettvangur",
              "target": "Vefforritari",
              "value": 1,
            },
            Object {
              "source": "Advania",
              "target": "Framendaforritari",
              "value": 1,
            },
            Object {
              "source": "Origo",
              "target": "Vefforritari",
              "value": 1,
            },
            Object {
              "source": "Vefforritari",
              "target": "Bið-athugun",
              "value": 1,
            },
            Object {
              "source": "Framendaforritari",
              "target": "Viðtal",
              "value": 1,
            },
            Object {
              "source": "Vefforritari",
              "target": "Viðtal",
              "value": 1,
            },
            Object {
              "source": "Viðtal",
              "target": "Annað-viðtal",
              "value": 1,
            },
            Object {
              "source": "Annað-viðtal",
              "target": "Neitun",
              "value": 1,
            },
          ],
          "nodes": Array [
            Object {
              "id": "Vettvangur",
              "label": "Vettvangur",
              "name": "Vettvangur",
            },
            Object {
              "id": "Advania",
              "label": "Advania",
              "name": "Advania",
            },
            Object {
              "id": "Origo",
              "label": "Origo",
              "name": "Origo",
            },
            Object {
              "id": "Vefforritari",
              "label": "Vefforritari",
              "name": "Vefforritari",
            },
            Object {
              "id": "Framendaforritari",
              "label": "Framendaforritari",
              "name": "Framendaforritari",
            },
            Object {
              "id": "Bið-athugun",
              "label": "Bið athugun",
              "name": "Bið-athugun",
            },
            Object {
              "id": "Viðtal",
              "label": "Viðtal",
              "name": "Viðtal",
            },
            Object {
              "id": "Annað-viðtal",
              "label": "Annað viðtal",
              "name": "Annað-viðtal",
            },
            Object {
              "id": "Neitun",
              "label": "Neitun",
              "name": "Neitun",
            },
          ],
        }
      }
      enableLabels={true}
      labelFormat={[Function]}
      labelOrientation="vertical"
      labelPadding={16}
      labelPosition="outside"
      labelTextColor="inherit:darker(1)"
      linkHoverOpacity={0.5}
      margin={
        Object {
          "bottom": 40,
          "left": 60,
          "right": 60,
          "top": 40,
        }
      }
      motionDamping={11}
      motionStiffness={120}
      nodeBorderColor="inherit:darker(0.8)"
      nodeOpacity={1}
      nodeWidth={18}
    />
  </div>
  <DateRange
    handler={[Function]}
    maxDate={2018-11-16T00:00:00.000Z}
    minDate={2018-11-10T00:00:00.000Z}
    selectionRange={
      Object {
        "endDate": 2018-11-16T00:00:00.000Z,
        "key": "selectionRange",
        "startDate": 2018-11-10T00:00:00.000Z,
      }
    }
  />
</div>
`;

exports[`Diagram component matches the snapshot without data 1`] = `
<div
  className="diagram-wrapper"
>
  <h2
    className="diagram__no-data"
  >
    Engar umsóknir
  </h2>
  <DateRange
    handler={[Function]}
    maxDate={Date { NaN }}
    minDate={Date { NaN }}
    selectionRange={
      Object {
        "endDate": Date { NaN },
        "key": "selectionRange",
        "startDate": Date { NaN },
      }
    }
  />
</div>
`;
